### PR TITLE
feat(kernel): annotate implemented interfaces on "ObjRef"s

### LINKED ISSUE #818: Allow "create" to specify additional interfaces
The kernel's `create` (and the private `objects.register`) call only requires the caller to specify a single `fqn` (that of the base type, which can be `Object` if nothing better is available). Unfortunately, this means that instances that implement one or more interface not inherited from a Javascript base class, the type information for interface members is lost when crossing the language boundary. This in turns causes all method arguments and return types to be inferred as `any`, further preventing type information from being available. This has been found to cause issues such as #807.

In order to address this, the kernel needs to allow the caller to also specify a list of implemented interfaces, so that it is able to access type information for members that are override in the foreign language. This is also useful when a client sub-classes a `jsii` type and adds new interface implementations to it, since that allows the type information for those to become available, too.

### PR DESCRIPTION
**Pull request by RomainMuller**
Annotating the interfaces implemented by an object instance with the
outbound Object Reference, and allowing the kernel's "create" call to
receive a list of interfaces allows the kernel to use more run-time type
information and operate safer.

Fixes #818

---

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license].

[Apache 2.0 license]: https://www.apache.org/licenses/LICENSE-2.0

### CODE REVIEW

File: packages/jsii-java-runtime/project/src/main/java/software/amazon/jsii/JsiiEngine.java [hunk 1]
```diff
@@ -198,6 +198,9 @@ public Object getObject(final JsonNode objRefNode) {
      * @return The Java class name.
      */
     private Class<?> resolveJavaClass(final String fqn) throws ClassNotFoundException {
+        if ("Object".equals(fqn)) {
```
Comment (rix0rrr) [inline packages/jsii-java-runtime/project/src/main/java/software/amazon/jsii/JsiiEngine.java]: This might be mentioned separately in the commit/merge message. Seems like a different fix.

File: packages/jsii-kernel/lib/api.ts [hunk 2]
```diff
@@ -75,13 +80,29 @@ export interface LoadResponse {
 }
 
 export interface CreateRequest {
+  /**
+   * The FQN of the class of which an instance is requested (or "Object")
+   */
   fqn: string;
+  /**
+   * The FQNs of interfaces the instance implements, if any. Declaring
```
Comment (rix0rrr) [inline packages/jsii-kernel/lib/api.ts]: > Declaring interfaces that the class denoted by `fqn` implements is not necessary.

Maybe describe these as "additional" interfaces then? In addition to the interfaces implemented by the class.

File: packages/jsii-kernel/lib/kernel.ts [hunk 3]
```diff
@@ -756,83 +746,90 @@ export class Kernel {
...
+        }
       }
     }
 
     return undefined;
   }
 
-  private _typeInfoForProperty(fqn: string, property: string): spec.Property {
-    const typeInfo = this._tryTypeInfoForProperty(fqn, property);
+  private _typeInfoForProperty(property: string, fqn: string, interfaces?: string[]): spec.Property {
```
Comment (rix0rrr) [inline packages/jsii-kernel/lib/kernel.ts]: In effect, shouldn't the last 2 arguments just be a list of FQNs, which can be either class or interface types?

File: packages/jsii-kernel/lib/objects.ts [hunk 4]
```diff
@@ -79,26 +95,37 @@ export class ObjectTable {
...
     }
 
     const existingRef = objectReference(obj);
     if (existingRef) {
+      if (interfaces) {
+        const allIfaces = new Set(interfaces);
+        for (const iface of existingRef[api.TOKEN_INTERFACES] || []) {
+          allIfaces.add(iface);
+        }
+        existingRef[api.TOKEN_INTERFACES] = this.removeRedundant(Array.from(allIfaces), fqn);
```
Comment (rix0rrr) [inline packages/jsii-kernel/lib/objects.ts]: What if the same anonymous object gets returned under 2 different interfaces? This code path will not be hit, right? Does that interface get 'added' to the list of interfaces for this object?

Comment (rix0rrr) [reply packages/jsii-kernel/lib/objects.ts]: On the other hand, `removeRedundant` feels fairly expensive, I wouldn't want to do that every time an object crosses the jsii boundary.

File: packages/jsii-kernel/lib/api.ts (hunk 2 above)
Comment (RomainMuller) [reply packages/jsii-kernel/lib/api.ts]: Yeah that's how I have started to name them. I'll have to go on the comments & make sure it's homogeneously used.

File: packages/jsii-kernel/lib/kernel.ts (hunk 3 above)
Comment (RomainMuller) [reply packages/jsii-kernel/lib/kernel.ts]: There can only be ONE class there, and I feel like it's easier to represent that information if it's a separate argument. I probably should rename it though.

File: packages/jsii-kernel/lib/objects.ts (hunk 4 above)
Comment (RomainMuller) [reply packages/jsii-kernel/lib/objects.ts]: This is still a WIP, but yeah, I have a local commit that also adds new interfaces (& removes redundant) when we return an object through a different "persona".

But I also worry about the intrinsic cost of doing this, so it might warrant a slightly deeper refactor to improve there...

**Comment by RomainMuller** [comment]
@rix0rrr  - annotating from the active side is needed as well. This is still a draft PR as I'm working through the front-ends now... I wanted to have it out there at least so I don't lose anything if my hard disk crashes 😅

File: packages/jsii-kernel/lib/api.ts [hunk 5]
```diff
@@ -75,13 +89,29 @@ export interface LoadResponse {
...
+   * declare members that are found in the `overrides` property.
+   */
+  interfaces?: string[];
+  /**
+   * Arguments to pass to the constructor of `fqn`. ("Object" accepts none)
+   */
   args?: any[];
+  /**
+   * Declarations of method overrides that should trigger callbacks
+   */
```
[the remark this answers is not in the document]
Comment (RomainMuller) [reply packages/jsii-kernel/lib/api.ts]: Yeah - I also prefer with a blank line but I got lazy on this file 😆

File: packages/jsii-kernel/lib/objects.ts [hunk 6]
```diff
@@ -132,9 +159,71 @@ export class ObjectTable {
...
+    if (ti.interfaces) {
+      for (const iface of ti.interfaces) {
+        if (this.interfaces.has(iface)) {
+          continue;
+        }
+        this.interfaces.add(iface);
+        this.addFromInterface(iface);
+      }
+    }
+
```
[the remark this answers is not in the document]
Comment (RomainMuller) [reply packages/jsii-kernel/lib/objects.ts]: This one was a merge artifact that I forgot to go fix...

**Comment by RomainMuller** [comment]
@bmaizels - yup, the code was already mostly covered, and the tests that broke were un-broken by the updated runtime code. I'll still have a second pass over that to make sure the particular edge case this aims to solve is... indeed covered.

**Comment by RomainMuller** [comment]
There was indeed a bug hiding in there (both in Java & .NET) due to insufficient test coverage of the feature. This is now fixed (including test)
